skill-tree /src/main/java/com/RDS/skilltree/Filters/JWTAuthenticationFilter.java

```diff
@@ -0,0 +1,80 @@
+package com.RDS.skilltree.Filters;
+
+import com.RDS.skilltree.Authentication.UserAuthenticationToken;
+import com.RDS.skilltree.User.*;
+import com.RDS.skilltree.utils.FetchAPI;
+import com.RDS.skilltree.utils.JWTUtils;
+import com.RDS.skilltree.utils.RDSUser.Response;
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.concurrent.CompletableFuture;
+
+public class JWTAuthenticationFilter extends OncePerRequestFilter {
+
+    @Autowired
+    private JWTUtils jwtUtils;
+
+    @Autowired
+    private FetchAPI fetchAPI;
+
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private UserRepository userRepository;
+
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+            throws ServletException, IOException {
+        String token = getJWTFromRequest(request);
+        try {
+            if (StringUtils.hasText(token) && jwtUtils.validateToken(token)) {
+                String rdsUserId = jwtUtils.getRDSUserId(token);
+
+
+                UserModel userModel;
+
+                if (userRepository.existsByRdsUserId(rdsUserId)) {
+                    userModel = userRepository.findByRdsUserId(rdsUserId).get();
+                } else {
+                    CompletableFuture<Response> userResponse = fetchAPI.getRDSUserData(rdsUserId);
+                    CompletableFuture.allOf(userResponse).join();
```

Comment:
Correct 💯 if the other service slows down, the Completable Future will become a blocking call untill the future gets the value.  This can lead the threads to be stuck and threads getting exhausted.

Reply:
Updated